plugins/toolkit/jetbrains-ultimate/src/software/aws/toolkits/jetbrains/services/rds/auth/IamAuth.kt

```diff
@@ -62,14 +82,44 @@ class IamAuth : DatabaseAuthProviderCompatabilityAdapter {
                 val credentials = getCredentials(connection)
                 DatabaseCredentialsAuthProvider.applyCredentials(connection, credentials, true)
             } catch (e: Throwable) {
-                result = Result.Failed
-                throw e
+                if (e is IllegalStateException && e.message?.contains("Token refresh started before session initialized") == true) {
+                    val simpleErrorInfo = SimpleErrorInfo(
+                        message("rds.validation.iam_sso_connection.error_info"),
+                        e,
+                        listOf(SsoNoTokenFix(project, connection))
+                    )
+                    throw KnownDatabaseException(simpleErrorInfo)
+                } else {
+                    result = Result.Failed
+                    throw e
+                }
             } finally {
                 RdsTelemetry.getCredentials(project = project, result = result, databaseCredentials = IAM, databaseEngine = connection.getDatabaseEngine())
             }
         }
     }
 
+    fun handleSsoAuthentication(project: Project, connection: ProtoConnection): ProtoConnection {
+        val authInformation = getAuthInformation(connection)
+        val profileCredentials =
+            CredentialManager.getInstance().getCredentialIdentifierById(authInformation.connectionSettings.credentials.id) as ProfileCredentialsIdentifierSso
+
+        val session = CredentialManager.getInstance()
+            .getSsoSessionIdentifiers()
+            .first { it.id == profileCredentials.sessionIdentifier }
+        val ssoConnection = ToolkitAuthManager.getInstance().getOrCreateSsoConnection(
+            UserConfigSsoSessionProfile(
+                configSessionName = profileCredentials.ssoSessionName,
+                ssoRegion = session.ssoRegion,
+                startUrl = session.startUrl,
+                scopes = session.scopes.toList()
+            )
+        )
+
+        reauthConnectionIfNeeded(project, ssoConnection)
+        return connection
+    }
+
```

Comment:
how can we reuse the existing logic instead of copy pasting it

Reply:
`ProfileCredentialsIdentifierSso` returns a user action that performs the authentication. We are returning/throwing an error with a fix to authenticate here. Only session and ssoConnection part is common. We will have to create a separate method for that.